tests/devices/qubit/test_dq_interpreter.py

```diff
@@ -0,0 +1,523 @@
+    assert qml.math.allclose(g, -jax.numpy.sin(0.5))
+
+
+def test_capture_remains_enabled_if_measurement_error():
+    """Test that capture remains enabled if there is a measurement error."""
+
+    @DefaultQubitInterpreter(num_wires=1, shots=None)
+    def g():
+        return qml.sample(wires=0)  # sampling with analytic execution.
+
+    with pytest.raises(NotImplementedError):
+        g()
+
+    assert qml.capture.enabled()
+
+
+def test_pytree_function_output():
+    """Test that the results respect the pytree output of the function."""
+
+    @DefaultQubitInterpreter(num_wires=1, shots=None)
+    def g():
+        return {
+            "probs": qml.probs(wires=0),
+            "state": qml.state(),
+            "var_Z": qml.var(qml.Z(0)),
+            "var_X": qml.var(qml.X(0)),
+        }
+
+    res = g()
+    assert qml.math.allclose(res["probs"], [1.0, 0.0])
+    assert qml.math.allclose(res["state"], [1.0, 0.0 + 0j])
+    assert qml.math.allclose(res["var_Z"], 0.0)
+    assert qml.math.allclose(res["var_X"], 1.0)
+
+
+def test_mcm_reset():
+    """Test that mid circuit measurements can reset the state."""
+
+    @DefaultQubitInterpreter(num_wires=1)
+    def f():
+        qml.X(0)
+        qml.measure(0, reset=True)
+        return qml.state()
+
+    out = f()
+    assert qml.math.allclose(out, jnp.array([1.0, 0.0]))  # reset into zero state.
+
+
+def test_operator_arithmetic():
+    """Test that dq can execute operator arithmetic."""
+
+    @DefaultQubitInterpreter(num_wires=2)
+    def f(x):
+        qml.RY(1.0, 0)
+        qml.adjoint(qml.RY(x, 0))
+        _ = qml.SX(1) ** 2
+        return qml.expval(qml.Z(0) + 2 * qml.Z(1))
+
+    output = f(0.5)
+    expected = jnp.cos(1 - 0.5) - 2 * 1
+    assert qml.math.allclose(output, expected)
+
+
+class TestSampling:
+    """Test cases for generating samples."""
+
+    def test_known_sampling(self):
+        """Test sampling output with deterministic sampling output"""
+
+        @DefaultQubitInterpreter(num_wires=2, shots=10)
+        def sampler():
+            qml.X(0)
+            return qml.sample(wires=(0, 1))
+
+        results = sampler()
+
+        expected0 = jax.numpy.ones((10,))  # zero wire
+        expected1 = jax.numpy.zeros((10,))  # one wire
+        expected = jax.numpy.vstack([expected0, expected1]).T
+
+        assert qml.math.allclose(results, expected)
+
+    def test_same_key_same_results(self):
+        """Test that two circuits with the same key give identical results."""
+        key = jax.random.PRNGKey(1234)
+
+        @DefaultQubitInterpreter(num_wires=1, shots=100, key=key)
+        def circuit1():
+            qml.Hadamard(0)
+            return qml.sample(wires=0)
+
+        @DefaultQubitInterpreter(num_wires=1, shots=100, key=key)
+        def circuit2():
+            qml.Hadamard(0)
+            return qml.sample(wires=0)
+
+        res1 = circuit1()
+        res2 = circuit2()
+
+        assert qml.math.allclose(res1, res2)
```

Comment:
To make sure that randomness is fixed beyond the first execution. (Might seem pedantic but i found the logic around `key` and `initial_key` complicated enough to make it worth testing this, imo)
```suggestion
        res1_first_exec = circuit1()
        res2_first_exec = circuit2()
        res1_second_exec = circuit1()
        res2_second_exec = circuit2()

        assert qml.math.allclose(res1_first_exec, res2_first_exec)
        assert qml.math.allclose(res1_second_exec, res2_second_exec)
```